### ISSUE TITLE
Add Limitation: overlay graph driver + linux kernel 4.14+

### ISSUE THREAD

**Issue by mcastelino** [clearcontainers/runtime#820]
# Description of problem

When using Clear Containers with overlay graph driver with linux kernel 4.14... causes multiple issues.
The file system accesses may hang. The system may report out of memory.

This seems to be an issues with overlay+9p+kernel 4.14. Until this issue has been fixed, Clear Container should not be run on a host with 4.14+ kernel and overlay graph driver.

**Comment by gtkramer** [comment]
This issue is also seen with overlay2 on Clear Linux

**Comment by rarindam** [comment]
Not able to see this issue with a plain qemu (2.9, 2.10 from qemu github) + kernel (2.14, 2.14-rc8), with sharing of tmp folder with lots of symlinks (host "etc" folder copied with links preserved, onto shared folder).
Before running any ls command, from root prompt of the guest, can you issue:
"tail -f /var/log/{messages,kernel,dmesg,syslog} &"
Then issue ls -l /root.
Hopefully this could capture some kernel logs.

**Comment by amshinde** [comment]
@rarindam I had tried this with qemu 2.7.1. I had seen no messages of interest in dmesg. Is the folder you are sharing using overlay file system?
 I can try this with qemu 2.9 and see if I can reproduce.

**Comment by sboeuf** [comment]
@rarindam I guess that you're sharing without relying on overlay, right ? The issue only happens if the shared filesystem relies on overlay.

**Comment by rarindam** [comment]
@sboeuf 
Yes exactly. I am not using docker. This is plain old Ubuntu host + qemu vm + Ubuntu guest with custom built kernel. The shared folder is copy of my host /etc with symlinks preserved.

**Comment by sboeuf** [comment]
@rarindam well in that case, no issue for us either (even with qemu 2.7). Could you try to create an overlay fs, pass this to your VM, and try again ? As @amshinde mentioned, I am curious if this is reproducible with qemu >= 2.9

**Comment by rarindam** [comment]
Is there anyway I can request a bzip2 file with the overlay filesystem which any of you are using?
Thanks ahead.

**Comment by amshinde** [comment]
@rarindam I just sent you a zip file. If you have docker installed you can get an overlay rootfs using this
```
# create the rootfs directory
mkdir rootfs

# export busybox via Docker into the rootfs directory
docker export $(docker create busybox) | tar -C rootfs -xvf -
```

**Comment by amshinde** [comment]
Just tried this with qemu 2.9. I see the same behaviour.

**Comment by sboeuf** [comment]
@amshinde thanks for the confirmation

**Comment by miguelinux** [comment]
The first commit which shows the issue is
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=b5efccbe0a12a9d5c65c1a60b4270837c7fdb900

**Comment by gtkramer** [comment]
@rarindam, @miguelinux has found the commit that changes how overlayfs functions which impacts QEMU.  I can grant you access to a system that has this issue for debugging.

**Comment by miguelinux** [comment]
@gtkramer @rarindam my bad that one is the last good one, the first bad is
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=4edb83bb1041e2f946ce36ea93f6bcd06d984bf4

**Comment by rarindam** [comment]
Hi Archana,
I unzipped the rootfs, and mounted the folder onto a virtual machine after booting.
I chrooted to the mounted folder, and issues ls –l commands on the different folders.
I don’t see a crash.
I am running qemu-2.9.0 (from https://github.com/qemu/qemu.git, not qemu_lite), and Kernel 4.14, and I don’t see this issue.
I am not much aware of the overlay system, so cannot pinpoint whether its due to overlay or not.
But certainly seems like a docker specific issue.
Arindam

**Comment by amshinde** [comment]
@miguelinux  Is there an bug open for the issue upstream?

**Comment by rarindam** [comment]
Hi Archana,
Lets have a debug session over skype.

But before that, CC is passing 9p related parameter at boot time to the CC kernel.
Please add debug=0x888 to that.
For more reference: https://www.kernel.org/doc/Documentation/filesystems/9p.txt.

Let me know if you are able to do that.
Arindam

**Comment by mcastelino** [comment]
@amshinde have we tried different caching settings

**Comment by amshinde** [comment]
@mcastelino Just tried with cache=mmap and fscache and loose.Does not help.

**Comment by amshinde** [comment]
@rarindam That debug option needs to be passed while mounting the 9p share right? I was able to do that.

**Comment by miguelinux** [comment]
@amshinde no issue open upstream.

Also I did a test with a VM and 9P + OverlayFS with QEMU 2.10.1 and 4.14 and no issues found. 

Could it be our QEMU-lite?

**Comment by miguelinux** [comment]
@amshinde I modified `configuration.toml` file to use QEMU 2.10.1.

The issue still there :-(

our Qemu-lite is not the guilty

**Comment by amshinde** [comment]
@rarindam  Tried with 4.14.1 kernel in the guest, still see the issue :(

**Comment by rarindam** [comment]
Hi Yang,
Pulling you in this email thread to keep everybody updated.

Using 4.14 as host, 4.14 as guest, qemu 2.9:
Using Archana’s instructions created overlay and put the rootfs, and mount in guest…see the issue.

Enabling debug mode while mount from guest, shows any ls command on mounted FS goes to infinite loop.
However, I read any specific file, directly under the mounted FS, no issues seems.

It also seems that the issue is at the top level mount folder only.
If I issue “ls -l tmp/rootfs/bin” or “ls -l tmp/rootfs/etc” there is no hang.

But If I issue ls –l tmp/rootfs, there is infinite loop.
It seems that OverlayFS is somehow sending its root level directory info different
                                        OR
9PFS was assuming something which is non standard.

Miguel also pointed out that the issue seems to happen after a commit in OverlayFS, between 4.13 and 4.14.
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=4edb83bb1041e2f946ce36ea93f6bcd06d984bf4

Thanks,
Arindam

**Comment by miguelinux** [comment]
@rarindam @amshinde @mcastelino 

A patch to fix this issue in 4.14 is at
https://marc.info/?l=linux-unionfs&m=151193367801434

This patch fix the issue on **OverlayFS** and **9P FS**.
BTW this issue was a bug in **OverlayFS** not from **CC**.
The reported email is at: https://marc.info/?l=linux-unionfs&m=151192916900486&w=2 

I already applied this patch to Clear Linux.

**Comment by rarindam** [comment]
Thanks for the fix Miguel.  That was blazing fast for a kernel fix.

**Comment by amshinde** [comment]
@miguelinux Thanks for the fix!
